### PR TITLE
GH-44846: [C++] Fix thread-unsafe access in ConcurrentQueue::UnsyncFront

### PR DESCRIPTION
**Pull request by pitrou**
### Rationale for this change

The `UnsyncFront` method claims that it can access a `std::deque`'s front element without synchronizing with another thread that would call `std::deque::push_back`, but `std::deque::front` can actually be implemented in terms of `std::deque::begin` while `std::deque::push_back` is specified to invalidate iterators.

In the end, `UnsyncFront` is concretely thread-unsafe even though it might ideally be thread-safe. This shows up as occasional Thread Sanitizer failures.

### What changes are included in this PR?

Replace `UnsyncFront` with a thread-safe `Front` method.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

No.

* GitHub Issue: #44846

### CODE REVIEW

**Comment by pitrou** [comment]
@github-actions crossbow submit [test-ubuntu-24.04-cpp-thread-sanitizer](https://github.com/ursacomputing/crossbow/actions/runs/12001035805/job/33450981551)

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

File: cpp/src/arrow/acero/concurrent_queue_internal.h [hunk 1]
```diff
@@ -65,17 +65,19 @@ class ConcurrentQueue {
...
-  // Un-synchronized access to front
-  // For this to be "safe":
-  // 1) the caller logically guarantees that queue is not empty
-  // 2) pop/try_pop cannot be called concurrently with this
-  const T& UnsyncFront() const { return queue_.front(); }
-
-  size_t UnsyncSize() const { return queue_.size(); }
+  const T& Front() const {
+    // Need to lock the queue because `front()` may be implemented in terms
+    // of `begin()`, which isn't safe with concurrent calls to e.g. `push()`.
```
Comment (mapleFU) [inline cpp/src/arrow/acero/concurrent_queue_internal.h:70]: is this `push` means `deque::push_back/front` or `ConcurrentQueue::Push()`?

Comment (pitrou) [reply cpp/src/arrow/acero/concurrent_queue_internal.h:70]: Here it means `std::queue::push` :) (which is concretely the same as `std::deque::push_back`)

File: cpp/src/arrow/acero/concurrent_queue_internal.h [hunk 2]
```diff
@@ -65,17 +65,19 @@ class ConcurrentQueue {
...
+    // (see GH-44846)
+    std::unique_lock<std::mutex> lock(mutex_);
+    return queue_.front();
```
Comment (westonpace) [inline cpp/src/arrow/acero/concurrent_queue_internal.h:73]: Is it safe to return a reference or does a copy need to be made?

**Review by westonpace — APPROVED** [review]
Thanks for tackling this.  Approve minus one question.

**Comment by mapleFU** [comment]
🤔 would `std::list` be ok here? Should we merge this or trying list since it would not invalidate anything ( but has less cache locality and requires more spaces)

File: cpp/src/arrow/acero/concurrent_queue_internal.h (hunk 1 above)
Comment (mapleFU) [reply cpp/src/arrow/acero/concurrent_queue_internal.h:70]: Ah, the original code is too optimistic. The deque might implemented as something like `std::vector<block*>` and `block*` for elements, the `front()` might visit `std::vector<block*>`  and `push_back` might enlarge it, so it really has problem here

**Comment by pitrou** [comment]
`std::list` will perform much more poorly. My intuition is that locking the `front()` access is better.

**Pull request merged**